### PR TITLE
Detect and Revert Duplicate License Terms

### LINKED ISSUE #41: Duplicate License Terms allowed on Registering Derivative IP
## Expected Behavior
Can only link to parents using unique License Terms. If the user mints two or more License Tokens of the same License Terms from the same parent, and attempts to link to that parent using all of the License Tokens, it should revert.

## Actual Behavior
Linking to a parent with two or more of the same License Tokens (different IDs, since mint amount > 1) is permitted.
1. Mint two or more license tokens of the same terms (e.g. Non-Commercial Social Remix) from a parent
2. Use all the license tokens from (1) to link a single IP to the parent from (1).

## Example
- IPA 1 is root
- IPA 2 is child of IPA 1
- IPA 3 is child of IPA 2
- All licenses used are commercial with 10% royalty

Below results are produced

**Case 1**: If IPA 2 is linked to IPA 1 using one license (and the same between IPA 2 and 3), then
- IPA 1 has 10% over IPA 2
- IPA 1 has 20% over IPA 3
- IPA 2 has 10% over IPA 3

**Case 2 (bug)**: If IPA 2 is linked to IPA 1 using two same licenses (and IPA 3 is linked to IPA 2 using one license), then
- IPA 1 has 20% over IPA 2
- IPA 1 has 30% over IPA 3
- IPA 2 has 10% over IPA 3

### PR DESCRIPTION
**Pull request by kingster-will**
This PR detects duplicate license terms when registering derivative and revert operations.

Closes #41

### CODE REVIEW

File: contracts/registries/LicenseRegistry.sol [hunk 1]
```diff
@@ -238,9 +238,13 @@ contract LicenseRegistry is ILicenseRegistry, AccessManagedUpgradeable, UUPSUpgr
 
         for (uint256 i = 0; i < parentIpIds.length; i++) {
             _verifyDerivativeFromParent(parentIpIds[i], childIpId, licenseTemplate, licenseTermsIds[i]);
-            $.parentIps[childIpId].add(parentIpIds[i]);
             $.childIps[parentIpIds[i]].add(childIpId);
-            $.attachedLicenseTerms[childIpId].add(licenseTermsIds[i]);
+            // determine if duplicate license terms
```
Comment (LeoHChen) [inline contracts/registries/LicenseRegistry.sol:242]: do you need to check the return value of line 241/242?

**Review by LeoHChen — COMMENTED** [review]
LGTM, are there any other place you need to check the return value of the .add function, which may be causing bugs?

**Review by jdubpark — APPROVED** [review]

File: contracts/registries/LicenseRegistry.sol (hunk 1 above)
Comment (kingster-will) [reply contracts/registries/LicenseRegistry.sol:242]: We don't need to, as when registering a derivative to parents, we only check the duplication of license terms of parents.

**Comment by kingster-will** [comment]
> LGTM, are there any other place you need to check the return value of the .add function, which may be causing bugs?

The return value is for indicating whether the item has already exist in the Set. As this is the only one function to register derivative relationship between child Ip and parent IP need to check duplicate. no other place need this check.

**Review by LeoHChen — APPROVED** [review]

**Pull request merged**